util/pull/pull.go

```diff
@@ -0,0 +1,471 @@
+package pull
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/diff"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker/schema1"
+	"github.com/containerd/containerd/rootfs"
+	ctdsnapshot "github.com/containerd/containerd/snapshots"
+	"github.com/moby/buildkit/snapshot"
+	"github.com/moby/buildkit/source"
+	"github.com/moby/buildkit/util/imageutil"
+	"github.com/moby/buildkit/util/progress"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type Puller struct {
+	Snapshotter  snapshot.Snapshotter
+	ContentStore content.Store
+	Applier      diff.Applier
+	Src          *source.ImageIdentifier
```

Comment:
This is probably better as a reference type to avoid dependency on `source`

Reply:
updated